text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+control.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `Drop` trait gains one new method, with a default implementation:
+
+```rust
+trait Drop {
+    fn drop(&mut self);
+
+    fn poll_drop_ready(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        Poll::Ready(())
+    }
+}
+```
+
+Like `Drop::drop`, it would not be possible to call `poll_drop_ready` on a
+type (users should use `mem::poll_drop_ready` discussed below instead).
+
+When a value is dropped inside of an async context (that is, an async function
+or block), it's `poll_drop_ready` method will be called repeatedly until it
+returns `Poll::Ready(())`. Then, its `drop` method will be called. This way,
+users can perform (or prepare to perform) nonblocking operations during
+destructors when they are called in an async context.
+
+It's important to note, however, that `poll_drop_ready` may be called even
+after it has returned `Poll::Ready(())`. This is different from the `drop`
+method, which is generally guaranteed to be called only once. Users
+implementing `poll_drop_ready` should take care to ensure that it has "fuse"
+semantics - that once it returns `Ready(())`, it continues to return that value
+thereafter.
+
+These additional APIs are also added to the `std::mem` module:
+
+```rust
+// an empty async function
+async fn drop_async<T>(to_drop: T) { }
+
+fn poll_drop_ready<T>(&mut self, cx: &mut Context<'_>) -> Poll<()>
+    where T: ?Sized
+{
+    // implemented through a lang item
+}
+```
+
+The `drop_async` function is analogous to the `drop` function, and is also
+added to the prelude as well. This function drops the value in an async
+context, guaranteeing that its `poll_drop_ready` method will be called.
+
+The `poll_drop_ready` function calls this value's "drop ready glue" - it calls
+`poll_drop_ready` on this value and also on all of its fields, recursively. The
+exact ordering and semantics of this glue are specified in the reference
+section.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Guarantees about when `poll_drop_ready` is called
+
+In general, users cannot assume that `poll_drop_ready` will ever be called,
+just as they cannot assume that destructors will run at all. However, they can
+assume less about `poll_drop_ready` than they can about `drop`, and that should
+be noted.
+
+In particular, users cannot safely assume that their values will be dropped in
+an async context, regardless of how they are intended to be used. It's this
+easy to supress the `poll_drop_ready` call of a value:
+
+```rust
+async fn ignore_poll_drop_ready<T>(x: T) {
+    // by passing `x` to `drop`, a non-async context, `poll_drop_ready` is
+    // never called when dropping `x`.
+    drop(x);
+}
+```
+
+However, we do guarantee that values dropped in an async context *will* have
+`poll_drop_ready` called, and we even guarantee the drop order between
```

Comment:
Would it be possible to explicitly add that `poll_drop_ready` may not be called, but when it's called it's in a loop *at least* until it returns `Poll::Ready(())` ?  
(Yes it's been already said above, but I feel it should also be in the "guarantees" section)